File: python/problems/pg42898/solution.py
"""
[PG] 42898 - 등굣길
https://programmers.co.kr/learn/courses/30/lessons/42898
난이도: lv3
태그: DP

시간복잡도: O(N^2)
공간복잡도: O(1)
"""


def solution(m: int, n: int, puddles: list[list[int]]) -> int:
    # 1. 초기화
    graph = [[0]*m for _ in range(n)]
    graph[0][0] = 1
    for puddle in puddles:
        x, y = puddle[0], puddle[1]
        graph[y-1][x-1] = -1
    for y in range(n):
        for x in range(m):
            if graph[y][x] == 0:
                up, left = 0, 0
                if x > 0 and graph[y][x - 1] != -1:
                    left = graph[y][x-1]
                if y > 0 and graph[y - 1][x] != -1:
                    up = graph[y-1][x]
                graph[y][x] += left + up
    return graph[n-1][m-1] % 1_000_000_007

File: go/problems/pg42898/solution.go
// Package pg42898 - [PG] 42898 - 등굣길
// https://programmers.co.kr/learn/courses/30/lessons/42898
// 난이도: lv3
// 태그: DP
//
// 시간복잡도: O(NM) - N: y축 길이, M: x축 길이
// 공간복잡도: O(NM) - DP 테이블 크기
package pg42898

func Solution(m int, n int, puddles [][]int) int {
	graph := make([][]int, n)
	reminder := 1_000_000_007
	for i := range graph {
		graph[i] = make([]int, m)
	}
	for i := range puddles {
		y := puddles[i][1]
		x := puddles[i][0]
		graph[y-1][x-1] = -1
	}
	graph[0][0] = 1
	for y := 0; y < n; y++ {
		for x := 0; x < m; x++ {
			if graph[y][x] == 0 {
				up := 0
				left := 0
				if x > 0 && graph[y][x-1] != -1 {
					left = graph[y][x-1]
				}
				if y > 0 && graph[y-1][x] != -1 {
					up = graph[y-1][x]
				}
				graph[y][x] += (up + left) % reminder
			}
		}
	}
	return graph[n-1][m-1] % reminder
}
